ext/bin/pdb-dataset

```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/env bash
+
+set -ueo pipefail
+
+usage() { echo 'Usage: PDBBOX=BOX_DIR pdb-dataset import --timeshift-to now DUMPFILE'; }
+
+misuse()
+{
+  usage 1>&2
+  exit 1
+}
+
+if test "$#" -lt 1 -o -z "$PDBBOX"; then
+  misuse
+fi
+
+test $# -gt 0 || misuse
+command="$1"
+shift 1
+
+case "$command" in
+  import)
+    timeshift=''
+    timeshiftsql=''
+    dumpfile=''
+
+    while test $# -gt 0; do
+      case "$1" in
+        --timeshift-to)
+          test $# -gt 0 || misuse
+          timeshift="$2"
+          shift 2
+          ;;
+        *)
+          if [ -z "$dumpfile" ]; then
+            dumpfile="$1"
+            shift
+          else
+            echo "Unknown argument: $1" 1>&2
+            misuse
+          fi
+      esac
+    done
+
+    if [ "$timeshift" = 'now' ]; then
+      timshiftsql="now()"
+    elif [[ $timeshift =~ ^[0-9]{4}-[0-9]{2}-[0-9]{2}$ ]]; then
+      timshiftsql="TIMESTAMP '${timeshift}'"
+    else
+      echo "Invalid --timeshift value: $timeshift" 1>&2
+      misuse
+    fi
+
+    echo "Checking for existing puppetdb database" 1>&2
+    db_intialized=$(psql "puppetdb" -t -c "SELECT 1 FROM information_schema.tables WHERE table_name = 'schema_migrations';")
+
+    if [ "$db_intialized" ]; then
+      echo 'puppetdb database already exists' 1>&2
+      exit 2
+    fi
+
+    echo "Restoring database from backup" 1>&2
+    pg_restore --role=postgres --no-owner --no-acl -d puppetdb "$dumpfile" || echo $?
+
+    db_exists=$(psql "puppetdb" -t -c "SELECT 1 FROM information_schema.tables WHERE table_name = 'schema_migrations';")
+
+    if [ "$db_exists" ]; then
+      echo 'Restore failed.' 1>&2
+      exit 2
+    fi
+
+    tmpdir="$(mktemp -d "pdb-dataset-XXXXXX")"
+    tmpdir="$(cd "$tmpdir" && pwd -P)"
+    trap "$(printf 'rm -rf %q' "$tmpdir")" EXIT
+
+cat > "$tmpdir/updatetime.sql" <<EOF
+BEGIN;
+DROP TABLE IF EXISTS max_report;
+
+SELECT max(producer_timestamp)
+INTO TEMPORARY TABLE max_report
+FROM reports;
+
+DROP TABLE IF EXISTS max_resource_event;
+
+SELECT max(timestamp)
+INTO TEMPORARY TABLE max_resource_event
+FROM resource_events;
+
+DROP TABLE IF EXISTS time_diff;
+
+SELECT (DATE_PART('day', ${timshiftsql} - (select max from max_report)) * 24 +
+        DATE_PART('hour', ${timshiftsql} - (select max from max_report))) * 60 +
+        DATE_PART('minute', ${timshiftsql} - (select max from max_report)) as minute_diff
+INTO TEMPORARY TABLE time_diff;
+
+DROP TABLE IF EXISTS resource_events_time_diff;
+
+SELECT (DATE_PART('day', ${timshiftsql} - (select max from max_resource_event)) * 24 +
+        DATE_PART('hour', ${timshiftsql} - (select max from max_resource_event))) * 60 +
+        DATE_PART('minute', ${timshiftsql} - (select max from max_resource_event)) as minute_diff
+INTO TEMPORARY TABLE resource_events_time_diff;
+
+UPDATE reports
+  SET producer_timestamp = producer_timestamp + ((select minute_diff from time_diff) * INTERVAL '1 minute'),
```

Comment:
I was able to get this running to the point where it restored a puppetdb database from a pg_dump and then attempted to run the `updatetime.sql` script on the restored data. Doing this hits an issue with the constraints on existing report partitions. 

This `updatetime.sql` script was written before we had partitioned `reports` and `resource_events` tables so we'll need to change how the updating of timestamps is handled because the existing constraints on each partition won't allow you to update a row in place and increase its timestamp outside of the range allowed in the partition. After the script errors the transaction updating all the timestamps rollsback and you end up with a restored db with old timestamps. See example constraint error below: 

```
ERROR:  new row for relation "reports_20210416z" violates check constraint "reports_20210416z_producer_timestamp_check"
DETAIL:  Failing row contains (22, \x62c7899ab6e6219bd2766bd053e4ca9fa8b8b905, 2bce4bec-ce8b-4d36-91d5-e204af508e98, host-4, 3.7.2, 4, a3adc55e-514f-4483-80d3-a258646ced2c, 2021-05-13 15:47:47.293-07, 2021-05-13 15:47:52.293-07, 2021-05-18 12:33:09.63-07, f, 1, 1, null, null, 2021-05-13 15:47:57.293-07, [{"name": "anchor", "value": 0.00046438, "category": "time"}, {"..., [{"file": null, "line": null, "tags": ["notice"], "time": "2015-..., [{"file": "/home/wyatt/.puppet/modules/concat/manifests/fragment..., 76e7a26a-f7a2-44bc-b0fd-8db6edd8ec9e, on_failure, 87bf8011-918c-40b3-aded-cee4ec9d5536, 1, f, t, null, agent).
```